middleware/proxy/dns.go

```diff
@@ -32,7 +35,22 @@ func (d *dnsEx) Protocol() string          { return "dns" }
 func (d *dnsEx) OnShutdown(p *Proxy) error { return nil }
 func (d *dnsEx) OnStartup(p *Proxy) error  { return nil }
 
+func (d *dnsEx) tap(co net.Conn, m *dns.Msg, isQuery bool) error {
+	if d.Taper == nil {
+		// nothing to tap to
+		return nil
+	}
+	dat := msg.Data{}
+	dat.Epoch()
```

Comment:
dat.Epoch() is weird. The Epoch() function should just return the `uint64`
~~~
dat := msg.Data{TimeSec: Epoch()}
~~~
Or make a `new.NewData()` is TimeSec is always set.